src/object.d

```diff
@@ -4881,3 +4881,108 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * Used by `__ArrayCast` to obtain a descriptive error message.
+ * It is a template so it can be used by `__ArrayCast` in -betterC
+ * builds.  It is separate from `__ArrayCast` to minimize code
+ * bloat.
+ * Params:
+ *  fromType = name of the type being cast from
+ *  fromSize = total size in bytes of the array being cast from
+ *  toType   = name of the type being cast o
+ *  toSize   = total size in bytes of the array being cast to
+ * Returns:
+ *  a descriptive error message describing the cast failure
+ */
+private string generateArrayCastErrorMessage()(string fromType, size_t fromSize, string toType, size_t toSize) @trusted
+{
+    import core.internal.string : unsignedToTempString;
+    import core.stdc.stdlib : alloca;
+
+    const(char)[][8] msgComponents =
+    [
+        "Cannot cast `"
+        , fromType
+        , "` to `"
+        , toType
+        , "`; an array of size "
+        , unsignedToTempString(fromSize)
+        , " does not align on an array of size "
+        , unsignedToTempString(toSize)
+    ];
+
+    // convert discontiguous `msgComponents` to contiguous string on the stack
+    size_t length = 0;
+    foreach(m ; msgComponents)
+        length += m.length;
+
+    auto msg = (cast(char*)alloca(length))[0 .. length];
+
+    size_t index = 0;
+    foreach(m ; msgComponents)
+        foreach(c; m)
+            msg[index++] = c;
+
+    return cast(string)msg;
+}
+
+@safe @nogc pure nothrow unittest
+{
+    string msg = generateArrayCastErrorMessage("int[]", 10, "long[]", 8);
+    assert(msg == "Cannot cast `int[]` to `long[]`; an array of size 10 does not align on an array of size 8");
+}
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
+{
+    const fromSize = from.length * TFrom.sizeof;
+    const toLength = fromSize / TTo.sizeof;
+
+    if ((fromSize % TTo.sizeof) != 0)
+    {
+        const errorMessage = generateArrayCastErrorMessage(TFrom.stringof, fromSize, TTo.stringof, toLength * TTo.sizeof);
+
+        // must be `false` or `0` to ensure the assertion remains
+        // even when compiling with `-release`
```

Comment:
That comment is a bit misleading to future readers as the `assert(0)` expression gets replaced by a HLT instruction in `-release` ...)